docs/cpu_templates/cpu-templates.md

```diff
@@ -0,0 +1,162 @@
+
+```bash
+curl --unix-socket /tmp/firecracker.socket -i  \
+  -X PUT 'http://localhost/machine-config' \
+  -H 'Accept: application/json'            \
+  -H 'Content-Type: application/json'      \
+  -d '{
+           "vcpu_count": 2,
+           "mem_size_mib": 1024,
+           "cpu_template": "T2CL"
+  }'
+```
+
+## Custom CPU templates
+
+Users can create their own CPU templates by creating a json file containing
+modifiers for CPUID, MSRs or ARM registers.
+
+Note: Creating custom CPU templates requires expert knowledge of
+CPU architectures. An inappropriate configuration may lead to guest crashes or
+making guests vulnerable to security attacks. Custom CPU templates must be
+tested thoroughly before use in production.
+
+In order to assist with creation and usage of CPU templates, there exists
+a CPU template helper tool.
+TODO: add a link to the CPU template helper tool when it is available.
+
+### Configuring custom CPU templates
+
+Configuration of a custom CPU template is performed via `/cpu-config` API endpoint.
+
+An example of configuring a custom CPU template on x86_64:
+
+```bash
+curl --unix-socket /tmp/firecracker.socket -i  \
+  -X PUT 'http://localhost/cpu-config' \
+  -H 'Accept: application/json'            \
+  -H 'Content-Type: application/json'      \
+  -d '{
+        "cpuid_modifiers": [
+         {
+            "leaf": "0x1",
+           "subleaf": "0x0",
+           "flags": 0,
+           "modifiers": [
+             {
+              "register": "eax",
+              "bitmap": "0bxxxx000000000011xx00011011110010"
+             }
+           ]
+         }
+       ],
+       "msr_modifiers": [
+         {
+           "addr": "0x10a",
+           "bitmap": "0b0000000000000000000000000000000000000000000000000000000000000000"
+         }
+       ]
+  }'
+```
+
+An example of configuring a custom CPU template on ARM:
+
+```bash
+curl --unix-socket /tmp/firecracker.socket -i  \
+  -X PUT 'http://localhost/cpu-config' \
+  -H 'Accept: application/json'            \
+  -H 'Content-Type: application/json'      \
+  -d '{
+        "reg_modifiers": [
+          {
+            "addr": "0x603000000013c020",
+            "bitmap": "0bxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx0000xxxxxxxxxxxx0000xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
+          }
+        ]
+  }'
+```
+
+### Custom CPU templates language schema
+
+The full description of the custom CPU templates language can be found
+[here](schema.json).
+
+#### Expansion of contracted bitmaps
+
+If a contracted version of a bitmap is given, for example, `0x101` where
+a 32-bit bitmap is expected, missing characters are implied to be `x`.
+
+#### Default values for CPUID leaves that are not supported by KVM
+
+If a custom CPU template makes modifications to a CPUID leaf that is
+not reported as supported by KVM via `KVM_GET_SUPPORTED_CPUID`,
+the default register values of such a leaf are assumed to be `0`.
```

Comment:
Is this behavior true? It seems to return error in this case.
https://github.com/firecracker-microvm/firecracker/blob/feature/cpu-templates/src/vmm/src/guest_config/x86_64/mod.rs#L84-L88

Reply:
You are probably right, it was an updated comment that I addressed. Removed.